### PR TITLE
Enable easy zoom/drag

### PR DESCRIPTION
**Pull request by logust79**
- Made a change to `renderer.js` to enable easy zoom/drag
- Added a zoom/drag example in readme.md

### CODE REVIEW

**Comment by mstefaniuk** [comment]
@logust79 It is very cool feature. You changed `dist/renderer.js` which is result of minimizing sources located in `src`. Modified line is on https://github.com/logust79/graph-viz-d3-js/blob/master/src/js/stage.js#L74.
To make controlled changes you need `node.js`. Check short contributor guide below:
```bash
> npm install
> npm install -g grunt-cli
> grunt build test
```
Build should be successful with all tests passing. Then make changes and launch:
```bash
> grunt all
```
It would build, test and create new dist (takes some time). I suppose there is few places where main `g` group is assumed directly below `svg` node so some tiny tweaks are needed. I hope it is not too much for the first time. It is really superb feature which I would expose on Graphviz fiddling website.

I would extend `README` with contributing guide in near future.

**Comment by logust79** [comment]
Ah, thanks for letting me know ;)

**Comment by logust79** [comment]
Just did it. Don't know how it looks...

**Comment by mstefaniuk** [comment]
It looks that all core sources has been changed. Now I have one thing that should be taken into consideration. Stage component has method `svg()` for exporting diagram source. Now, when diagram can be zoomed and virtually cropped, it should grab whole unmodified diagram (without zooming rectangle) or it should grab only visible part (also without rectangle). Both are valuable.
The former can be obtained by removing additional `g` and by resetting translation / scale. Another way to handle it smoothly is to add zooming feature into `stage.js` (as an option) and export svg according to existence of this option. 
There is one naive test case in `stage-spec.js` for this function.

**Comment by logust79** [comment]
I actually just changed that line in stage.js. The rest of the changes are related to tests.

In terms of export, removing rectangle is easy. The additional g has the translation / scale fields, so removing the g will simply remove the zoom / drag with it. Removing the additional g might not be straightforward, involving copying large amount of innerHTML. Reseting the translation / scale fields could be an easy option.

I have a stupid question, how can I call the svg() function to export the svg code?

**Comment by mstefaniuk** [comment]
To be honest `stage.js` returns svg document for `renderer.js` `getImage()` function. It generates PNG file with diagram to be saved. Of course both functions has corresponding unit tests.

**Comment by logust79** [comment]
hmmm, I see. 

Maybe we can pass an optional parameter to the getImage function to ask if I like to save the zoomed / unzoomed image ?

I’ll play around. Thanks a lot for the explanation.

**Comment by mstefaniuk** [comment]
Optional boolean parameter in `getImage(bool cropToZoom)` would be perfect and backward compatible. Second thing is to add optional object parameter to render / stage `init()` and put there definition of zoom (size / position).
Is it possible to leave size to be defined in CSS (auto adjust for resolution)?

**Comment by logust79** [comment]
What I’ve been thinking is to pass data like this to init:
{element: element, zoomable: bool, zoomfunc: somefunc }

And in init, if zoomable, we provide a default zoom function if ‘somefunc’ is not provided.

**Comment by mstefaniuk** [comment]
What about size / position values - should these not be fixed? Currently there are pluggable transition functions (in stage) so these can be provided using this object.

So it init is called with string it would be an element selector. If there would be an object it would be extended options (with zooming).

**Comment by logust79** [comment]
Sorry I’m a bit confused. What do size / position do here? Can you be elaborate on this a little bit? Treat me as a noob ;)

**Comment by mstefaniuk** [comment]
In your snippet you included in `README` and I suppose it would be a part of stage initialization, there are following lines:
```javascript
    var svg = d3.select("div#graph").select('svg').select('g')
        .call(d3.behavior.zoom().scaleExtent([0.1, 10.0]).on('zoom', zoom));
    svg.select('g').append('rect')
        .attr('class', 'overlay')
        .attr('width', 3000)
        .attr('height', 1000)
        .attr('x', 0)
        .attr('y', -800);
    function zoom() {
        svg.select('g').attr('transform', 'translate(' + d3.event.translate + ')scale(' + d3.event.scale + ')');
    }
```
There is zoom function attached to the behavior - it is planned to be overriden. There is also width / height as well x / y so I assume it also should be configurable?

**Comment by logust79** [comment]
Oh, I now see what you mean. Yes the rect’s configuration can be calculated from the size of the graph. As long as it covers the whole region, it will catch all mouse zoom / drag events. I can also incorporate the overlay css in the d3 script, so users won’t need to configure it in the css.

**Comment by mstefaniuk** [comment]
There is already place to inline default SVG styles in [stage initialization](https://github.com/mstefaniuk/graph-viz-d3-js/blob/master/src/js/stage.js#L70).

**Comment by mstefaniuk** [comment]
@logust97 Do you plan to add any tests for new features?

**Comment by logust79** [comment]
Just managed to put everything in place ( apart from tests). One caveat is I have to let `renderer.init()` return a `d3.behavior.zoom()` instance for it to reset properly before saving the image.

**Comment by logust79** [comment]
I’ll try to add some tests in. But can’t guarantee the time.

File: src/js/renderer.js [hunk 1]
```diff
@@ -31,8 +31,11 @@ define(["stage", "worker!layout-worker.js"], function(stage, worker) {
         pending = source;
       }
     },
-    getImage: function() {
-      var svgXml = stage.svg();
+    getImage: function(obj) {
```
Comment (mstefaniuk) [inline src/js/renderer.js]: This method defaults export options to `zoom: 0` but `stage.svg()` accepts `reset: true`.

File: src/js/stage.js [hunk 2]
```diff
@@ -64,16 +64,48 @@ define(["d3", "palette", "transitions/default"], function (d3, palette, defaults
...
+				main.attr('transform', 'translate(' + d3.event.translate + ')scale(' + d3.event.scale + ')');
+			}
+			return zoom;
+        }
       },
-      svg: function () {
+      svg: function (obj) {
+      	if (typeof(obj) == 'object' && obj.reset) {
+      		var g = svg.select("g").select("g");
+      		if (g[0]){
```
Comment (mstefaniuk) [inline src/js/stage.js]: Side effect of resetting zoom for export would be zoom reset also for showed graph itself.

**Comment by mstefaniuk** [comment]
I don't get why you need to return zoom behavior from stage init. Why it cannot be hold inside stage and used within `svg()` function?

I would merge this request just after Easter holidays even if you would not commit any tests. Then we decide what to do.

**Comment by logust79** [comment]
Maybe I was thinking something not necessary. When asked to export the image, it resets the original as well. I guess this is unnecessary.

**Comment by logust79** [comment]
The reason why it returns the zoom instance is for reseting, which requires it in order to reset properly.

I’m not very sure now if the side effect as resetting is a good thing. What do you think? I would personally like to reset it anyway.

**Comment by mstefaniuk** [comment]
To leave diagram untouched it would need to clone, reset, export and then destroy it. May be improved later.

I realized one thing more. There is some transitions on svg document so when you reset a zoom it may take some time to make it happen actually. To make it simple it should not be animated at all.

**Comment by logust79** [comment]
I thought I removed all the transitions when return the svg to make sure it returns the right thing. Can you point me to the code?

**Comment by mstefaniuk** [comment]
You are right. There is no transitions attached to zoom `g`.

**Comment by logust79** [comment]
I have to confess I have never written a test in this fashion before. But when I write a little test to check whether stage.init({element:’#element’}) returns the right thing, it throws out an error like this:
TypeError: d3.behavior is undefined in /Users/logust/git/graph-viz-d3-js/src/js/stage.js (line 9)
Can you help me with this?

Many thanks,
Jing

**Comment by mstefaniuk** [comment]
I suppose you are trying to add a test in `stage-spec.js` which uses mocked and spied `d3.js` implementation. To do it this way you need to extend already [defined mock](https://github.com/mstefaniuk/graph-viz-d3-js/blob/master/spec/stage-spec.js#L86).

You can do it also without mocking in `stage-dom-spec.js` and define assertions based on resulting DOM structure. You may need to move `stage.init()` from `beforeEach()` to every test to make different calls possible.

**Comment by logust79** [comment]
Brilliant! It works by moving the test to stage-dom-spec.js.

Many thanks,
Jing

**Comment by mstefaniuk** [comment]
One more advice - if you comment on an issue / pull request please cut my comments off to make it easier to read on mobile.

**Comment by logust79** [comment]
Hi,
Just ask, how do I disable the initial transition?

Kind regards,
Jing

**Comment by mstefaniuk** [comment]
Within `stage-dom-spec.js` all transitions should be disabled using `shortcircut` function and provided into `stage` as `immediate` structure. Maybe there is a transition in zoom behavior?

**Comment by mstefaniuk** [comment]
Do you plan to add anything to this pull request? Could you check also issues [I pointed some time ago](https://github.com/mstefaniuk/graph-viz-d3-js/pull/17#issuecomment-199193808)?

**Comment by logust79** [comment]
Well spotted.. I will change it to reset: false

By the way, when I try to add some hover behaviours on the nodes, some of the nodes are covered by the arrows (class:relation). This is because class:relation are drawn after the class:node. In the stage.js, changing the group.sort order fix this problem.

groups.sort(function (a, b) {
        //nodes draw after relations, 
          return order[a.class] - order[b.class];  #<<- change it to `return order[b.class] - order[a.class]`
        });

I can add this change if you prefer.

Jing

**Comment by mstefaniuk** [comment]
I saw that you are working on another change (not zooming) but I would like to keep them separated (as you have with branches). If you would like to change shapes sorting you should rather change [order definition in `stage.js`](/mstefaniuk/graph-viz-d3-js/blob/master/src/js/stage.js#L3) but not sorting function.

**Comment by logust79** [comment]
Oh, this is cool!

Thanks for letting me know this.

Jing

**Comment by logust79** [comment]
I thought I separated all the non-zoom behaviour changes to the branch. Can you help point out the irrelevant changes on my master?

Kind regards,
Jing

**Comment by mstefaniuk** [comment]
You are right - code is separated but your comment was not.

**Comment by logust79** [comment]
Haha, sorry for the confusion. Not very good at commenting :)

**Pull request merged**